azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -969,223 +930,336 @@ def do_retry():
-                # we can distinguish between connect ops that we're handling (they go into the
-                # queue) and connect ops that we are sending down.
-                #
-                # Once we finally connect, we only have to complete the ops in the queue and we
-                # never have to worry about completing the op that we sent down.  The code is much
-                # cleaner this way, especially when you take retries into account, trust me.
-                self.waiting_connect_ops.append(op)
-                self._send_new_connect_op_down()
-
-        elif isinstance(op, pipeline_ops_base.DisconnectOperation):
-            if self.state == ReconnectState.WAITING_TO_RECONNECT:
-                logger.info(
-                    "{}({}): State changes {}->LOGICALLY_DISCONNECTED.  Canceling waiting ops and sending disconnect down.".format(
-                        self.name, op.name, self.state
-                    )
-                )
-                self.state = ReconnectState.LOGICALLY_DISCONNECTED
-                self._clear_reconnect_timer()
-                self._complete_waiting_connect_ops(
-                    pipeline_exceptions.OperationCancelled("Explicit disconnect invoked")
-                )
-                # TODO: There is a bug here. If in a WAITING_TO_RECONNECT state, and an explicit
-                # Disconnect occurs, it is completed here rather than going all the way down. This
-                # means that any in-flight operations are NOT cancelled.
-                op.complete()
+                self.waiting_ops.put_nowait(op)
+                self.state_lock.release()
 
             else:
-                logger.info(
-                    "{}({}): State changes {}->LOGICALLY_DISCONNECTED.  Sending op down.".format(
-                        self.name, op.name, self.state
-                    )
-                )
-                self.state = ReconnectState.LOGICALLY_DISCONNECTED
-                self.send_op_down(op)
+                if isinstance(op, pipeline_ops_base.ConnectOperation):
+                    if self.state is ReconnectState.CONNECTED:
+                        logger.debug(
+                            "{}({}): State is already CONNECTED. Sending op down".format(
+                                self.name, op.name
+                            )
+                        )
+                        self._add_connection_op_callback(op)
+                    elif self.state is ReconnectState.DISCONNECTED:
+                        logger.debug(
+                            "{}({}): State changes DISCONNECTED -> CONNECTING. Sending op down".format(
+                                self.name, op.name
+                            )
+                        )
+                        self.state = ReconnectState.CONNECTING
+                        self._add_connection_op_callback(op)
+                    else:
+                        # This should be impossible to reach
```

Comment:
># This should be impossible to reach [](http://example.com/codeflow?start=24&length=36)

I don't think this is true -- what prevents this from happening?  We could be in ReconnectState.CONNECTING or ReconnectState.DISCONNECTING for a long time.

Reply:
The block above adds any ops to `waiting_ops` in the case where we are in a "progress state", i.e. CONNECTING, DISCONNECTING, REAUTHORIZING.

This entire section is an else block, indicating that we already know we are not in a "progress state"

Reply:
duh.  I looked back for that before commenting and missed it.   I recomment at least adding that as a comment.  Making it a warning made me think it was a "should not happen, but no big deal if it does".  It feels more like an assert because hitting this line indicates a funamental flaw in the code.  (I wish we had those).

Reply:
I could switch it to a logger.error rather than a warning?

Reply:
it couldn't hurt.  Or a logger.critical.